File: Ass_3_6.java
import java.util.*;

public class Ass_3_6 {

    // Helper function to generate all subset sums for a part of the array
    private static ArrayList<Long> getSubsetSums(long[] arr) {
        ArrayList<Long> sums = new ArrayList<>();
        sums.add(0L); // Start with an empty subset sum of 0

        for (long num : arr) {
            int currentSize = sums.size();
            // For every existing sum, add the current number to create new sums
            for (int i = 0; i < currentSize; i++) {
                sums.add(sums.get(i) + num);
            }
        }
        return sums;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Check if input exists to avoid crashes
        if (!sc.hasNext()) {
            sc.close();
            return;
        }

        int T = sc.nextInt(); // Number of test cases

        while (T-- > 0) {
            int N = sc.nextInt();
            long S = sc.nextLong(); // Target sum

            long[] A = new long[N];
            for (int i = 0; i < N; i++) {
                A[i] = sc.nextLong();
            }

            // --- Meet-in-the-Middle Logic ---

            // 1. Split array into two halves
            int mid = N / 2;
            long[] leftPart = Arrays.copyOfRange(A, 0, mid);
            long[] rightPart = Arrays.copyOfRange(A, mid, N);

            // 2. Generate all subset sums for both halves
            ArrayList<Long> leftSums = getSubsetSums(leftPart);
            ArrayList<Long> rightSums = getSubsetSums(rightPart);

            // 3. Sort the right half sums to enable Binary Search
            Collections.sort(rightSums);

            // 4. Find the closest sum
            long minDiff = Long.MAX_VALUE;

            for (long u : leftSums) {
                long target = S - u;

                // Binary search for 'target' in rightSums
                int idx = Collections.binarySearch(rightSums, target);

                if (idx >= 0) {
                    // Exact match found! Difference is 0.
                    minDiff = 0;
                    break;
                } else {
                    // Not found, convert return value to insertion point
                    int insertionPoint = -(idx + 1);

                    // Check the element at insertionPoint (value > target)
                    if (insertionPoint < rightSums.size()) {
                        long diff = Math.abs((u + rightSums.get(insertionPoint)) - S);
                        if (diff < minDiff) minDiff = diff;
                    }

                    // Check the element before insertionPoint (value < target)
                    if (insertionPoint > 0) {
                        long diff = Math.abs((u + rightSums.get(insertionPoint - 1)) - S);
                        if (diff < minDiff) minDiff = diff;
                    }
                }
            }

            System.out.println(minDiff);
        }
        
        sc.close();
    }
}

File: Ass_3_6.py
import sys
from bisect import bisect_left

def get_subset_sums(arr):
    """
    Generates all possible subset sums for a list of integers.
    """
    sums = [0]
    for num in arr:
        # Append (current_sum + num) to the list
        # We use a fixed range to avoid infinite loops while modifying the list
        current_len = len(sums)
        for i in range(current_len):
            sums.append(sums[i] + num)
    return sums

def solve():
    # Read all input from standard input
    input_data = sys.stdin.read().split()
    
    if not input_data:
        return

    iterator = iter(input_data)
    
    try:
        t_str = next(iterator, None)
        if not t_str: return
        T = int(t_str)
        
        for _ in range(T):
            # Read N and Target S
            N = int(next(iterator))
            S = int(next(iterator))
            
            A = []
            for _ in range(N):
                A.append(int(next(iterator)))
            
            # --- Meet-in-the-Middle Logic ---
            
            # 1. Split array into two halves
            mid = N // 2
            left_part = A[:mid]
            right_part = A[mid:]
            
            # 2. Generate all subset sums for both halves
            left_sums = get_subset_sums(left_part)
            right_sums = get_subset_sums(right_part)
            
            # 3. Sort the right half sums to enable Binary Search
            right_sums.sort()
            
            # 4. Find the closest sum
            min_diff = float('inf')
            
            # For each sum 'u' in the left half, we want a 'v' in the right half
            # such that u + v is closest to S.
            # Ideally, v should be close to (S - u).
            
            right_len = len(right_sums)
            
            for u in left_sums:
                target = S - u
                
                # Find the insertion point where 'target' fits in sorted right_sums
                idx = bisect_left(right_sums, target)
                
                # Check the value at the insertion index (if within bounds)
                # This value is >= target
                if idx < right_len:
                    diff = abs((u + right_sums[idx]) - S)
                    if diff < min_diff:
                        min_diff = diff
                
                # Check the value immediately before the insertion index (if within bounds)
                # This value is < target
                if idx > 0:
                    diff = abs((u + right_sums[idx - 1]) - S)
                    if diff < min_diff:
                        min_diff = diff
                
                # Optimization: If we found an exact match (diff is 0), we can't do better
                if min_diff == 0:
                    break
            
            print(min_diff)

    except StopIteration:
        pass

if __name__ == "__main__":
    solve()